### PR TITLE
Fail if meta is not a pandas object

### LINKED ISSUE #8537: dask apply loses index type with simple meta
**What happened**:
When trying to *apply* a function to a dataframe, and giving a simple meta, the perceived dtype of the index of the return value is lost. Also the name of the index is lost (harder to show in this minimal example, but pretty similar I assume).
In the bigger picture this gives issues with ddf.merge where it expects the index column "[0]" to exist but instead finds "nameofcolumn".

**What you expected to happen**:
The perceived dtype of the index of the return value is the same as the dtype of the index of return value when computed.

**Minimal Complete Verifiable Example**:

image

```python
import dask.dataframe as dd
from dask.datasets import timeseries
ddf = timeseries()
ddf.index.dtype                        # shows dtype('<M8[ns]')
editedx = ddf.x.apply(lambda x: x + 1, meta=ddf.x.dtype) # some simple example, giving a simple meta argument
editedx.head()
editedx.index.dtype                  # incorrectly shows dtype('int64'), the default
editedx.index.compute().dtype # correctly shows dtype('<M8[ns]'), as was the dtype of the initial data

```

**Anything else we need to know?**:
It works in this case when we leave the meta out, but in some cases the autodeduction that happens is wrong and we need to provide a specific (if sometimes simple) meta to fix this).

**Environment**:

- Dask version: 2021.12.0
- Python version: 3.9.7
- Operating System: linux (jupyterhub docker on k8s)
- Install method (conda, pip, source): (installed with the base jupyterhub image from above, source I guess? pip freeze shows dask @ file:///home/conda/feedstock_root/build_artifacts/dask-core_1639176667673/work)

### PR DESCRIPTION
**Pull request by jsignell**
- [x] Closes #8537
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

Still need to add a test to prove it.

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -6061,9 +6063,16 @@ def map_partitions(
...
+        if meta_is_emulated:
+            # If `meta` is not a pandas object, the concatenated results will be a
+            # different type
+            meta = make_meta(_concat([meta]), index=meta_index)
+        else:
+            raise ValueError(
+                "Meta is not valid, `map_partitions` expects output to be a pandas object. "
+                "Try passing a pandas object as meta or a dict or tuple representing the "
+                "(name, dtype) of the columns."
+            )
```
Comment (scharlottej13) [inline dask/dataframe/core.py:6075]: I have a couple questions to check my understanding!

1. if `meta_is_emulated is False & (has_parallel_type(meta) or is_arraylike(meta) and meta.shape)`, then we know `meta` is not valid? And that's why there's no need for an `elif meta_is_emulated is False` condition?
2. It seems there were already helpful errors raised for when inferring `meta` doesn't work if the user doesn't pass anything (with [`dataframe.utils.raise_on_meta_error`](https://github.com/dask/dask/blob/c1c88f066672c0b216fc24862a2b36a0a9fb4e22/dask/dataframe/utils.py#L165-L196)), but less so for when the user passes in something invalid?

**Review by scharlottej13 — COMMENTED** [review]
Just added a couple clarification questions!

File: dask/dataframe/core.py (hunk 1 above)
Comment (jsignell) [reply dask/dataframe/core.py:6075]: 1. the `else` here is coming from `if meta_is_emulated` so it is equivalent to `elif meta_is_emulated is False`. Basically I am trying to say, if the meta is not a pandas object but dask tried to guess the meta (`meta_is_emulated is True`) then let's fix it. If on the other hand the meta is not a pandas object and the user provided it explicitly (`meta_is_emulated is False`) then raise an error.
2. Right. If you look at the original issue you can see the kind of confusing behavior that you get. Essentially the meta was reduced down to a single value (it was `1`) at this spot in the codebase, and then it was being coerced back into a series, so the index was totally lost.

Comment (scharlottej13) [reply dask/dataframe/core.py:6075]: Ahh that makes sense, thank you! The other question I thought of is whether this is an issue elsewhere, but I think I've answered that one well enough for myself, though maybe not very efficiently. I went through the other methods with a `meta` argument and it seems like they either have a more strict enforcement of the meta type (e.g. in [`Series.map`](https://github.com/dask/dask/blob/c1c88f066672c0b216fc24862a2b36a0a9fb4e22/dask/dataframe/core.py#L3602) or rely on `map_partitions` (which will now be improved!)

**Review by scharlottej13 — APPROVED** [review]

**Comment by gjoseph92** [comment]
FYI I believe this is failing [`distributed/tests/test_collections.py::test_dataframe_groupby_tasks`](https://github.com/dask/distributed/blob/d956ef6b7ae09634fe9307e9da1e2de1fc487e8d/distributed/tests/test_collections.py#L184): https://github.com/dask/distributed/runs/4927021633?check_suite_focus=true#step:12:1679

**Comment by jsignell** [comment]
ooh thanks for mentioning that. I will open a PR on distributed.

**Pull request merged**